### PR TITLE
Show draft programs in API Docs

### LINKED ISSUE #4945: Show draft program spec in docs
Expand the program-specific API docs to also show the the expected response for the draft version of a program.

### PR DESCRIPTION
**Pull request by michaelzetune**
### Description

This PR adds on an additional dropdown in the API docs for the version of a program you want to view: Active or Draft. If that version is not available, we show a simple error message.

## Release notes

Adds version selection to API docs.

#### General

Read the full guidelines for PRs [here](https://docs.civiform.us/contributor-guide/developer-guide/technical-contribution-guide#guidelines-for-pull-requests)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Followed steps to [internationalize new strings](https://docs.civiform.us/contributor-guide/developer-guide/internationalization-i18n#internationalization-for-application-strings)
- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Tested on mobile view. See [mobile device mode](https://developer.chrome.com/docs/devtools/device-mode/)
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order

### Issue(s) this completes

Relates to #5262, fixes #4945.

### CODE REVIEW

File: browser-test/src/api_docs.test.ts [hunk 1]
```diff
@@ -9,20 +9,64 @@ import {BASE_URL} from './support/config'
...
+    await page.goto(BASE_URL)
+    await loginAsAdmin(page)
+
+    await adminPrograms.publishAllDrafts()
+    await page.click('text=API docs')
+
+    await validateScreenshot(page, 'comprehensive-program-active-version')
+
+    await page.selectOption('#select-slug', {value: 'minimal-sample-program'})
+    await validateScreenshot(page, 'minimal-program-active-version')
```
Comment (bion) [inline browser-test/src/api_docs.test.ts:31]: **suggestion (blocking):** in cases where the test manipulates the UI and then does a screenshot assertion, please insert an assertion on the updated text content of the page before doing the screenshot validation. This avoids a race condition where if the server is too slow to update the screenshot will be of the pre-updated UI.

File: server/app/views/api/ApiDocsView.java [hunk 2]
```diff
@@ -83,31 +84,50 @@ public Content render(
...
       Http.Request request) {
 
     SelectTag slugsDropdown =
         select()
             .withId("select-slug")
-            .withClasses("border", "border-gray-300", "rounded-lg", "p-2", "ml-2")
-            .attr("onchange", "window.location.href = '/api/docs/v1/' + this.value");
-
+            .withClasses("border", "border-gray-300", "rounded-lg", "p-2", "mx-2")
+            .attr("onchange", "window.location.href = '/api/docs/v1/' + this.value + '/active'");
```
Comment (bion) [inline server/app/views/api/ApiDocsView.java:104]: **suggestion (blocking):** please move this code into a typescript file and follow the pattern used for registering DOM event handlers used on other pages

**Review by bion — CHANGES_REQUESTED** [review]
Please add unit tests for the code changes in `VersionRepository` and `ProgramServiceImpl`

File: server/app/views/api/ApiDocsView.java (hunk 2 above)
Comment (michaelzetune) [reply server/app/views/api/ApiDocsView.java]: Done.

File: server/app/views/api/ApiDocsView.java [hunk 3]
```diff
@@ -83,31 +84,50 @@ public Content render(
...
+                    .withValue("active")
+                    .withCondSelected(request.uri().contains("/active")))
+            .with(
+                option("Draft")
+                    .withValue("draft")
+                    .withCondSelected(request.uri().contains("/draft")))
+            .attr(
+                "onchange",
+                "window.location.href = '/api/docs/v1/' +"
+                    + " document.getElementById('select-slug').value + '/' + this.value");
```
[the remark this answers is not in the document]
Comment (michaelzetune) [reply server/app/views/api/ApiDocsView.java]: Done.

File: server/app/assets/javascripts/api_docs.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import {addEventListenerToElements} from './util'
+
+class ApiDocs {
+  // Redirects the user to the active version of the program slug they just selected upon
+  // clicking the select program slug.
+  static attachSlugEventListener() {
+    addEventListenerToElements('#select-slug', 'change', (event: Event) => {
+      console.log('test log')
```
Comment (bion) [inline server/app/assets/javascripts/api_docs.ts:8]: **suggestion (blocking):** remove console log

File: server/app/repository/ProgramRepository.java [hunk 5]
```diff
@@ -173,6 +173,32 @@ public CompletableFuture<Program> getActiveProgramFromSlug(String slug) {
...
   }
 
+  /** Get the current draft program with the provided slug. */
+  public CompletableFuture<Program> getDraftProgramFromSlug(String slug) {
+    return supplyAsync(
+        () -> {
+          for (Program program : database.find(Program.class).where().isNull("slug").findList()) {
+            program.getSlug();
+            program.save();
+          }
```
Comment (bion) [inline server/app/repository/ProgramRepository.java:183]: **issue (blocking):** This is unrelated to the name and javadoc of the method. In general reads should not alter database state unless they're explicitly named `findOrCreate`. If it's necessary for the feature to work please refactor for clarity, otherwise remove.

File: server/app/controllers/api/ApiDocsController.java [hunk 6]
```diff
@@ -34,32 +36,49 @@ public ApiDocsController(
...
 
-    ProgramDefinition programDefinition;
     try {
-      programDefinition =
-          programService.getActiveProgramDefinitionAsync(programSlug).toCompletableFuture().join();
+      ProgramDefinition programDefinition =
+          getProgramDefFn.apply(programSlug).toCompletableFuture().join();
+      return Optional.of(programDefinition);
+
     } catch (Exception e) {
```
Comment (bion) [inline server/app/controllers/api/ApiDocsController.java:80]: **suggestion (blocking):** I know you didn't add this line, but we should **never** catch `Exception` like this, please update to a more specific exception type.

File: server/app/repository/ProgramRepository.java [hunk 7]
```diff
@@ -173,6 +173,32 @@ public CompletableFuture<Program> getActiveProgramFromSlug(String slug) {
...
+
+          Optional<Version> version = versionRepository.get().getDraftVersion();
+
+          if (version.isEmpty()) {
+            throw new RuntimeException(new ProgramNotFoundException(slug));
```
Comment (bion) [inline server/app/repository/ProgramRepository.java:188]: **suggestion (blocking):** 

1. would prefer this throws a checked exception so the caller decides how to handle it rather than the repository deciding for the caller that this will cause a server error.

2. use a different exception type -- `ProgramNotFoundException` doesn't match the exceptional case here that there is no draft version.

File: server/app/repository/ProgramRepository.java [hunk 8]
```diff
@@ -173,6 +173,32 @@ public CompletableFuture<Program> getActiveProgramFromSlug(String slug) {
...
+          }
+
+          ImmutableList<Program> draftPrograms = version.get().getPrograms();
+          List<Program> programsMatchingSlug =
+              database.find(Program.class).where().eq("slug", slug).findList();
+          return draftPrograms.stream()
+              .filter(programsMatchingSlug::contains)
+              .findFirst()
+              .orElseThrow(() -> new RuntimeException(new ProgramNotFoundException(slug)));
```
Comment (bion) [inline server/app/repository/ProgramRepository.java:197]: **suggestion (blocking):** throw the checked exception

File: server/app/repository/ProgramRepository.java [hunk 9]
```diff
@@ -173,6 +173,32 @@ public CompletableFuture<Program> getActiveProgramFromSlug(String slug) {
...
+
+          Optional<Version> version = versionRepository.get().getDraftVersion();
+
+          if (version.isEmpty()) {
+            throw new RuntimeException(new ProgramNotFoundException(slug));
+          }
+
+          ImmutableList<Program> draftPrograms = version.get().getPrograms();
+          List<Program> programsMatchingSlug =
+              database.find(Program.class).where().eq("slug", slug).findList();
```
Comment (bion) [inline server/app/repository/ProgramRepository.java:193]: **suggestion (blocking):** I don't think this second query is necessary? Looks like you could loop through `draftPrograms` and find the one with the matching `slug`

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/repository/ProgramRepository.java (hunk 7 above)
Comment (michaelzetune) [reply server/app/repository/ProgramRepository.java:181]: For (1), this is tricky because Java does not allow checked exceptions in CompletableFutures, even if we declare `throws ProgramNotFoundException` in the method signature (since the method itself isn't throwing it). That's probably why the existing `getActiveProgramFromSlug` above is wrapping it in a `RuntimeException` as well.

For this I could make this method synchronous (avoiding the supplyAsync block) since there isn't any async code calling it right now, or we could leave the unchecked exception as-is. What do you think?

For (2), I could still make the underlying `ProgramNotFoundException` a different one, perhaps a new `ProgramDraftNotAvailableException`?

File: server/app/repository/ProgramRepository.java (hunk 8 above)
Comment (michaelzetune) [reply server/app/repository/ProgramRepository.java:189]: Same question as above.

File: server/app/controllers/api/ApiDocsController.java (hunk 6 above)
Comment (michaelzetune) [reply server/app/controllers/api/ApiDocsController.java]: Updated to RuntimeException. I'll leave this comment unresolved until we figure out the best approach to throwing in the other method, since this might change.

Comment (michaelzetune) [reply server/app/controllers/api/ApiDocsController.java]: https://github.com/civiform/civiform/pull/5315 😄

File: server/app/repository/ProgramRepository.java (hunk 5 above)
Comment (bion) [reply server/app/repository/ProgramRepository.java]: 🙈 that is not going to make it easier for me to sleep this weekend. Thanks for pointing that out.

File: server/app/controllers/api/ApiDocsController.java (hunk 6 above)
Comment (bion) [reply server/app/controllers/api/ApiDocsController.java]: Thanks!

File: server/app/repository/ProgramRepository.java (hunk 5 above)
Comment (bion) [reply server/app/repository/ProgramRepository.java]: Looking at Program and ProgramDefinition, I think this was added originally because the `slug` column wasn't populated for all programs at one point, so this code was included as a hacky way of doing a data migration. All new programs have the column populated though and it's been a while, so I think it's safe to remove this code.

File: server/app/repository/ProgramRepository.java (hunk 7 above)
Comment (bion) [reply server/app/repository/ProgramRepository.java:181]: > For this I could make this method synchronous (avoiding the supplyAsync block) since there isn't any async code calling it right now, or we could leave the unchecked exception as-is. What do you think?

I'd prefer this -- better to avoid the complexity of async when we don't need it.

Comment (bion) [reply server/app/repository/ProgramRepository.java:181]: For (2): the important thing is to communicate to callers what could go wrong and what actually went wrong when it goes wrong. The problem wasn't that they asked for a program that doesn't exist, it's that they should not call this method when no draft version exists, so the exception and javadoc should reflect that.

File: server/app/repository/ProgramRepository.java (hunk 5 above)
Comment (michaelzetune) [reply server/app/repository/ProgramRepository.java]: I'll do this in a separate PR so that it's easy to roll back if necessary.

File: server/app/repository/ProgramRepository.java (hunk 7 above)
Comment (michaelzetune) [reply server/app/repository/ProgramRepository.java]: Made `getDraftProgramFromSlug` and `getDraftProgramDefinition` synchronous. Created and used new `ProgramDraftNotFoundException`.

File: server/app/repository/ProgramRepository.java (hunk 8 above)
Comment (michaelzetune) [reply server/app/repository/ProgramRepository.java]: `ProgramDraftNotFoundException` is now thrown

File: server/app/repository/ProgramRepository.java (hunk 5 above)
Comment (michaelzetune) [reply server/app/repository/ProgramRepository.java]: https://github.com/civiform/civiform/pull/5321

**Review by bion — APPROVED** [review]
